core/jvm/src/test/scala/fs2/ConcurrentlySpec.scala

```diff
@@ -0,0 +1,37 @@
+package fs2
+
+import scala.concurrent.duration._
+import cats.effect.IO
+
+class ConcurrentlySpec extends Fs2Spec {
+
+  "concurrently" - {
+
+    "when background stream terminates, overall stream continues" in forAll { (s1: PureStream[Int], s2: PureStream[Int]) =>
+      runLog(Scheduler[IO](1).flatMap(scheduler => (scheduler.sleep_[IO](25.millis) ++ s1.get).concurrently(s2.get))) shouldBe s1.get.toVector
+    }
+
+    "when background stream fails, overall stream fails" in forAll { (s: PureStream[Int], f: Failure) =>
+      val prg = Scheduler[IO](1).flatMap(scheduler => (scheduler.sleep_[IO](25.millis) ++ s.get).concurrently(f.get))
+      val throws = f.get.drain.run.attempt.unsafeRunSync.isLeft
+      if (throws) an[Err.type] should be thrownBy runLog(prg)
+      else runLog(prg)
+    }
+
+    "when primary stream fails, overall stream fails and background stream is terminated" in forAll { (f: Failure) =>
+      var bgDone = false
+      val bg = Stream.repeatEval(IO(1)).onFinalize(IO(bgDone = true))
+      val prg = Scheduler[IO](1).flatMap(scheduler => (scheduler.sleep_[IO](25.millis) ++ f.get).concurrently(bg))
+      an[Err.type] should be thrownBy runLog(prg)
+      bgDone shouldBe true
+    }
+
+    "when primary stream termiantes, background stream is terminated" in forAll { (s: PureStream[Int]) =>
+      var bgDone = false
+      val bg = Stream.repeatEval(IO(1)).onFinalize(IO(bgDone = true))
+      val prg = Scheduler[IO](1).flatMap(scheduler => (scheduler.sleep_[IO](25.millis) ++ s.get).concurrently(bg))
+      runLog(prg)
+      bgDone shouldBe true
+    }
+  }
```

Comment:
@mpilquist in 0.9 it won't work. in 0.8 however, that will work. My concern here is that if we won't be able to somehow resolve this, then this `concurrently` pattern is inherently unsafe and will lead to unsound dead-locked programs. Also, current mergeDrainL, R, etc patterns are suffering from this, and as such migration of complex concurrents programs from 0.8 to 0.9 is really painful :-(. 

@djspiewak, @pchiusano  I would be really grateful on your input here, I think it is really important design decision we are going to make here, and hence your worked on many concurrent primitives, perhaps you may share your opinion on this ?